core/trino-main/src/test/java/io/trino/sql/planner/optimizations/TestAddExchangesPlans.java

```diff
@@ -978,6 +990,226 @@ public void testJoinNotExactlyPartitioned()
+                                .right(
+                                        exchange(LOCAL, GATHER, SINGLE_DISTRIBUTION,
+                                                exchange(REMOTE, REPLICATE, FIXED_BROADCAST_DISTRIBUTION,
+                                                        project(
+                                                                tableScan("region", ImmutableMap.of("regionkey", "regionkey")))))))));
+    }
+
+    @Test
+    public void testUnionAllAboveBroadcastJoin()
+    {
+        assertDistributedPlan(
+                """
+                    SELECT regionkey FROM nation UNION ALL (SELECT nationkey FROM nation n JOIN region r on r.regionkey = n.nationkey)
+                """,
+                noJoinReordering(),
+                anyTree(
+                        exchange(REMOTE, GATHER, SINGLE_DISTRIBUTION,
+                                tableScan("nation"),
+                                join(INNER, join -> join
+                                        .equiCriteria("nationkey", "regionkey")
+                                        .left(
+                                                project(
+                                                        node(FilterNode.class,
+                                                                tableScan("nation", ImmutableMap.of("nationkey", "nationkey")))))
+                                        .right(
+                                                exchange(LOCAL, GATHER, SINGLE_DISTRIBUTION,
+                                                        exchange(REMOTE, REPLICATE, FIXED_BROADCAST_DISTRIBUTION,
+                                                                project(
+                                                                        tableScan("region", ImmutableMap.of("regionkey", "regionkey"))))))))));
+    }
+
+    @Test
+    public void testGroupedAggregationAboveUnionAllCrossJoined()
+    {
+        assertDistributedPlan(
+                """
+                    SELECT sum(nationkey) FROM (SELECT nationkey FROM nation UNION ALL SELECT nationkey FROM nation), region group by nationkey
+                """,
+                noJoinReordering(),
+                anyTree(
+                        join(INNER, join -> join
+                                .left(
+                                        exchange(LOCAL, REPARTITION, FIXED_ARBITRARY_DISTRIBUTION,
+                                                tableScan("nation", ImmutableMap.of("nationkey", "nationkey")),
+                                                tableScan("nation")))
+                                .right(
+                                        exchange(LOCAL, GATHER, SINGLE_DISTRIBUTION,
+                                                exchange(REMOTE, REPLICATE, FIXED_BROADCAST_DISTRIBUTION,
+                                                        tableScan("region")))))));
+    }
+
+    @Test
+    public void testGroupedAggregationAboveUnionAll()
```

Comment:
maybe `union all + table_scan + values`, `union all + table_scan + hash aggregation` if there arent' there?

Reply:
> union all + table_scan + hash aggregation

It is covered by `io.trino.sql.planner.optimizations.TestAddExchangesPlans#testUnionAllOnSourceAndHashDistributedChildren`

> `union all + table_scan + values`

added